include/deal.II/fe/fe_tools_extrapolate.templates.h

```diff
@@ -1164,94 +1183,92 @@ namespace FETools
+          const std::vector<std::vector<char>> array_of_strings =
+            Utilities::unpack<std::vector<std::vector<char>>>(buffer_recv);
+
+          for (const auto &s : array_of_strings)
+            {
+              CellData cell_data;
+              cell_data.unpack_data(s);
+              cell_data.receiver = other_rank;
 
-      const unsigned int n_senders =
-        Utilities::MPI::compute_n_point_to_point_communications(communicator,
-                                                                destinations);
+              received_cells.emplace_back(std::move(cell_data));
+            }
 
-      // receive data
-      std::vector<char> receive;
-      CellData          cell_data;
-      for (unsigned int index = 0; index < n_senders; ++index)
-        {
-          MPI_Status status;
-          int ierr = MPI_Probe(MPI_ANY_SOURCE, mpi_tag, communicator, &status);
-          AssertThrowMPI(ierr);
-
-          int len;
-          ierr = MPI_Get_count(&status, MPI_BYTE, &len);
-          AssertThrowMPI(ierr);
-          receive.resize(len);
-
-          char *buf = receive.data();
-          ierr      = MPI_Recv(buf,
-                          len,
-                          MPI_BYTE,
-                          status.MPI_SOURCE,
-                          status.MPI_TAG,
-                          communicator,
-                          &status);
-          AssertThrowMPI(ierr);
-
-          cell_data.unpack_data(receive);
-
-          // this process has to send this
-          // cell back to the sender
-          // the receiver is the old sender
-          cell_data.receiver = status.MPI_SOURCE;
-
-          received_cells.push_back(cell_data);
-        }
+          // Nothing left to do here, we don't actually need to provide an
+          // answer:
+          request_buffer.clear();
+        };
+
+      const auto read_answer = [](const unsigned int /*other_rank*/,
+                                  const std::vector<char> &recv_buffer) {
+        // We don't put anything into the answers, so nothing should
+        // have been coming out at this end either:
+        (void)recv_buffer;
+        Assert(recv_buffer.size() == 0, ExcInternalError());
+      };
 
-      if (requests.size() > 0)
-        {
-          const int ierr =
-            MPI_Waitall(requests.size(), requests.data(), MPI_STATUSES_IGNORE);
-          AssertThrowMPI(ierr);
-        }
+      Utilities::MPI::ConsensusAlgorithms::AnonymousProcess<char, char>
+        operations(get_destinations,
+                   create_request,
+                   answer_request,
+                   prepare_buffer_for_answer,
+                   read_answer);
 
-      // finally sort the list of cells
-      std::sort(received_cells.begin(), received_cells.end());
+      Utilities::MPI::ConsensusAlgorithms::NBX<char, char>(operations,
+                                                           communicator)
```

Comment:
I wouldn't use NBX directly but he selector. It also picks a specialized path for serial simulations.

Reply:
Ah yes, good suggestion. I had not learned about `Selector` when I wrote this part of the patch :-)